Generalise associative operator parsing

nagisa opened the pull request:
This commit generalises parsing of associative operators from left-associative
only (with some ugly hacks to support right-associative assignment) to properly
left/right-associative operators.

Parsing is still is not general enough to handle non-associative,
non-highest-precedence prefix or non-highest-precedence
postfix operators (e.g. `..` range syntax) and should be made to be.

Lastly, this commit adds support for parsing right-associative `<-` (left arrow)
operator with precedence higher than assignment as the operator for placement-in
feature.

---

This PR still needs various non-parser changes (e.g. src/grammar and tests) and I’m still working on these; the meat of the PR can already be reviewed, though, I think.

Please review carefully. I made sure that quirks I have discovered so far are preserved (see e.g. https://github.com/rust-lang/rust/issues/29071) and am looking for more corner cases as I continue to work on tests et al, but there may be something I haven’t noticed or accounted for.

EDIT: I’m also not sure I managed to preserve all the semantics with the range operator inside non-trivial expressions since these are a mess at the moment. Crater runs would be nice.

rust-highfive commented:
r? @brson

(rust_highfive has picked a reviewer for you, use r? to override)

nagisa commented:
Something I found is that `x = 2..3 = y` now parses as `(x = 2..3) = y` while it used to fail before. The expression still fails to compile later with `E0070` anyway, though.

Something that doesn’t parse anymore and is a bug is `x = ..y` and `x = ..`. I’ll tackle these soon.

brson commented:
r? @pnkfelix

nagisa commented:
Something that also needs to be fixed:

```
test.rs:45:5: 45:13 warning: unnecessary parentheses around assigned value, #[warn(unused_parens)] on by default
test.rs:45     (x <- y) <- z;
               ^~~~~~~~
```

alexcrichton commented:
Running a crater build to evaluate these changes

pnkfelix commented:
> Something that doesn’t parse anymore and is a bug is x = ..y and x = ... I’ll tackle these soon.

@nagisa So this is now fixed with your latest commit ?

nagisa commented:
Į believe it is.
On Oct 16, 2015 1:25 PM, "Felix S Klock II" <notifications@github.com>
wrote:

> Something that doesn’t parse anymore and is a bug is x = ..y and x = ...
> I’ll tackle these soon.
>
> @nagisa <https://github.com/nagisa> So this is now fixed with your latest
> commit ?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/rust-lang/rust/pull/29072#issuecomment-148677166>.
>

alexcrichton commented:
Crater reports [two regressions](https://gist.github.com/alexcrichton/343c7b21811126567457) one of which I believe is spurious and one of which I believe is legitimate.

nagisa commented:
Thanks, @alexcrichton!

nagisa commented:
This PR is now in a pretty good spot, I think. The underlying issue behind the regression was just me forgetting to move restriction filtering over from the old code. This also helped to fix issues I had with semi-statement things I hacked around previously.

pnkfelix commented:
(i won't get to this review until monday; hopefully that's okay.)

nagisa commented on src/libsyntax/parse/parser.rs:
```diff
@@ -2703,86 +2779,24 @@ impl<'a> Parser<'a> {
+        try!(self.bump());
+        let opt_end = if self.is_at_start_of_range_notation_rhs() {
+            // RHS must be parsed with more associativity than DotDot.
+            let next_prec = AssocOp::from_token(&token::DotDot).unwrap().precedence() + 1;
```
NB: this changes the previous bahaviour of `..` from 

> This has the same precedence as assignment expressions
> (much lower than other prefix expressions) to be consistent
> // with the postfix-form 'expr..'”

To having more precedence than `<-` and `=` but still less than any other binary operator.

nagisa commented:
NB: this still needs changes to `src/grammar` and I had planned to add more tests for quirky cases such as `x..y=z` not parsing (it did not parse before, and AFAIR it does not parse after changes).

Got distracted by university things, sadly.

nagisa commented:
I pushed `src/grammar` change, but honestly have no idea what I’m doing since I’m just following grammar for similarly behaving equalities.

Going from master to after the change, make check-grammar reports 2 less mismatches (from 132 down to 130).

pnkfelix commented:
> EDIT: ah also, the in PLACE {EXPR} warning is removed in nagisa@58c299f

Oh, okay, sorry.  Sometimes the commit-by-commit reviewing strategy works well, but I need to learn to look ahead in the queue for things like this.

nagisa commented:
> Oh, okay, sorry. Sometimes the commit-by-commit reviewing strategy works well, but I need to learn to look ahead in the queue for things like this.

I removed it after your comment, just made a note that I fixed it.

pnkfelix commented:
lgtm

pnkfelix commented:
@bors r+

bors commented:
:pushpin: Commit 99f9bb1 has been approved by `pnkfelix`

<!-- @bors r=pnkfelix 99f9bb16ab83519d67deaf4ca8a78feeba67fb53 -->

bors commented:
:hourglass: Testing commit 99f9bb1 with merge e0e2627...

bors commented:
:sunny: Test successful - [auto-linux-32-nopt-t](http://buildbot.rust-lang.org/builders/auto-linux-32-nopt-t/builds/6907), [auto-linux-32-opt](http://buildbot.rust-lang.org/builders/auto-linux-32-opt/builds/6921), [auto-linux-64-debug-opt](http://buildbot.rust-lang.org/builders/auto-linux-64-debug-opt/builds/180), [auto-linux-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-linux-64-nopt-t/builds/6860), [auto-linux-64-opt](http://buildbot.rust-lang.org/builders/auto-linux-64-opt/builds/6883), [auto-linux-64-x-android-t](http://buildbot.rust-lang.org/builders/auto-linux-64-x-android-t/builds/6866), [auto-linux-cross-opt](http://buildbot.rust-lang.org/builders/auto-linux-cross-opt/builds/102), [auto-linux-musl-64-opt](http://buildbot.rust-lang.org/builders/auto-linux-musl-64-opt/builds/2052), [auto-mac-32-opt](http://buildbot.rust-lang.org/builders/auto-mac-32-opt/builds/6937), [auto-mac-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-mac-64-nopt-t/builds/6933), [auto-mac-64-opt](http://buildbot.rust-lang.org/builders/auto-mac-64-opt/builds/6897), [auto-win-gnu-32-nopt-t](http://buildbot.rust-lang.org/builders/auto-win-gnu-32-nopt-t/builds/1888), [auto-win-gnu-32-opt](http://buildbot.rust-lang.org/builders/auto-win-gnu-32-opt/builds/1889), [auto-win-gnu-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-win-gnu-64-nopt-t/builds/1890), [auto-win-gnu-64-opt](http://buildbot.rust-lang.org/builders/auto-win-gnu-64-opt/builds/1888), [auto-win-msvc-32-opt](http://buildbot.rust-lang.org/builders/auto-win-msvc-32-opt/builds/1153), [auto-win-msvc-64-opt](http://buildbot.rust-lang.org/builders/auto-win-msvc-64-opt/builds/1872)
